mlflow/pyfunc/__init__.py

```diff
@@ -2981,32 +2982,46 @@ def predict(self, context, model_input: List[str], params=None) -> List[str]:
         if callable(python_model):
             # first argument is the model input
             type_hints = _extract_type_hints(python_model, input_arg_index=0)
-            signature_from_type_hints = _infer_signature_from_type_hints(
-                func=python_model, type_hints=type_hints, input_example=input_example
-            )
+            if _should_infer_signature_from_type_hints(type_hints):
+                signature_from_type_hints = _infer_signature_from_type_hints(
+                    func=python_model, type_hints=type_hints, input_example=input_example
+                )
         elif isinstance(python_model, PythonModel):
             saved_example = _save_example(mlflow_model, input_example, path, example_no_conversion)
-            signature_from_type_hints = _infer_signature_from_type_hints(
-                func=python_model.predict,
-                type_hints=python_model._get_type_hints(),
-                input_example=input_example,
-            )
+            type_hints = python_model._get_type_hints()
+            infer_signature_from_type_hints = _should_infer_signature_from_type_hints(type_hints)
+            if infer_signature_from_type_hints:
+                signature_from_type_hints = _infer_signature_from_type_hints(
+                    func=python_model.predict,
+                    type_hints=python_model._get_type_hints(),
+                    input_example=input_example,
+                )
             # only infer signature based on input example when signature
             # and type hints are not provided
-            if (
-                signature is None
-                and signature_from_type_hints is None
-                and saved_example is not None
-            ):
-                try:
-                    context = PythonModelContext(artifacts, model_config)
-                    python_model.load_context(context)
-                    mlflow_model.signature = _infer_signature_from_input_example(
-                        saved_example,
-                        _PythonModelPyfuncWrapper(python_model, None, None),
+            if signature is None and signature_from_type_hints is None:
+                warning_msg = (
```

Comment:
Can we simply define error messages for different cases one by one, rather than trying to combining sentences? The resulting message might sound unnatural, for example:

```
Type hint `xyz` cannot be used to infer model signature. Failed to infer model signature from input example, error: xyz
```
This looks like the inferring signature from input example is failed due to type hint and I'd get confused what to fix. Instead I think we should have a clear message like "Failed to infer model signature from input example or type hint. "